test/test_propagate_DWT.py

```diff
@@ -0,0 +1,219 @@
+
+            coeffs, Ucoeffs, ol = wave_dec(x, Ux, ld, hd)
+
+            #for c, Uc in zip(coeffs, Ucoeffs):
+            #    print(c)
+            #    print("-"*60)
+            #print("="*60)
+
+            # compare to the output of PyWavelet
+            result_pywt = pywt.wavedec(x, filter_name, mode='constant')
+
+            # compare output depth
+            assert len(result_pywt) == len(coeffs)
+            
+            # compare output in detail
+            for a, b in zip(result_pywt, coeffs):
+                assert len(a) == len(b)
+                assert np.allclose(a, b)
+
+
+def test_decomposition_realtime():
+    for filter_name in ["db2", "db3"]:
+
+        for nx in [20,21]:
+
+            x = np.random.randn(nx)
+            Ux = 0.1 * (1 + np.random.random(nx))
+
+            ld, hd, lr, hr = filter_design(filter_name)
+
+            # run x all at once
+            coeffs_a, Ucoeffs_a, ol_a, z_a = wave_dec_realtime(x, Ux, ld, hd, n=2)
+
+            # slice x into smaller chunks and process them in batches
+            # this tests the internal state options
+            coeffs_list = []
+            Ucoeffs_list = []
+            z_b = None
+            n_splits = 3
+            for x_batch, Ux_batch in zip(np.array_split(x, n_splits), np.array_split(Ux, n_splits)):
+                coeffs_b, Ucoeffs_b, ol_b, z_b = wave_dec_realtime(x_batch, Ux_batch, ld, hd, n=2, level_states=z_b)
+                coeffs_list.append(coeffs_b)
+                Ucoeffs_list.append(Ucoeffs_b)
+
+            coeffs_b = [np.concatenate([coeffs[level] for coeffs in coeffs_list], axis=0) for level in range(len(coeffs_list[0]))]
+            Ucoeffs_b = [np.concatenate([Ucoeffs[level] for Ucoeffs in Ucoeffs_list], axis=0) for level in range(len(Ucoeffs_list[0]))]
+
+            # compare output depth
+            assert len(coeffs_a) == len(coeffs_b)
+            
+            # compare output in detail
+            for a, b in zip(coeffs_a, coeffs_b):
+                assert len(a) == len(b)
+                assert np.allclose(a, b)
+
+
+def test_reconstruction():
+    pass
+
+
+def test_identity_multi():
+
+    for filter_name in ["db3", "db4"]:
+
+        for nx in [20, 21, 203]:
+
+            x = np.linspace(1,nx,nx)  # np.random.randn(nx)
+            #Ux = 0.1 * (1 + np.random.random(nx))
+            Ux = np.ones((nx))
+            Ux[nx//2:] = 2
+            Ux = 0.1 * Ux
+
+            ld, hd, lr, hr = filter_design(filter_name)
+
+            # full decomposition
+            coeffs, Ucoeffs, ol = wave_dec(x, Ux, ld, hd, kind="diag")
+
+            # full reconstruction
+            xr, Uxr = wave_rec(coeffs, Ucoeffs, lr, hr, original_length=ol, kind="diag")
+
+            assert x.size == xr.size
+            assert np.allclose(x, xr)
+            assert Ux.size == Uxr.size
+
+            #if nx == 203 and filter_name == "db4":
+            #    plt.plot(Ux)
+            #    plt.plot(Uxr)
+            #    plt.show()
```

Comment:
Same as above... delete it or add an explanatory comment would be my suggestion.

Reply:
:+1: